libraries/c_sdk/standard/https/src/iot_https_client.c

```diff
@@ -0,0 +1,3326 @@
+                if( HTTPS_FAILED( scheduleStatus ) )
+                {
+                    IotLogError( "Error scheduling HTTPS request %d. Error code: %d", pNextHttpsRequest, scheduleStatus );
+
+                    if( pNextHttpsRequest->isAsync && pNextHttpsRequest->pCallbacks->errorCallback )
+                    {
+                        pNextHttpsRequest->pCallbacks->errorCallback( pNextHttpsRequest->pUserPrivData, pNextHttpsRequest, NULL, scheduleStatus );
+                    }
+                    else
+                    {
+                        pNextHttpsRequest->pHttpsResponse->syncStatus = scheduleStatus;
+                    }
+                }
+            }
+        }
+        else
+        {
+            IotLogDebug( "Network receive callback found the request queue empty. A network send task was not scheduled." );
+        }
+    }
+
+    /* Dequeue response from the response queue now that it is finished. */
+    IotMutex_Lock( &( pHttpsConnection->respQMutex ) );
+
+    /* There could be a scenario where the request fails to send and the network server still responds,
+     * In this case, the failed response will have been cancelled and removed from the queue. If the network
+     * server still got a response, then the safest way to remove the current response is to remove it explicitly
+     * from the queue instead of dequeuing the header of the queue which might not be the current response. */
+    if( IotLink_IsLinked( &( pCurrentHttpsResponse->link ) ) )
+    {
+        IotDeQueue_Remove( &( pCurrentHttpsResponse->link ) );
+    }
+
+    IotMutex_Unlock( &( pHttpsConnection->respQMutex ) );
+
+    /* Signal to a synchronous reponse that the response is complete. */
+    if( pCurrentHttpsResponse->isAsync && pCurrentHttpsResponse->pCallbacks->responseCompleteCallback )
+    {
+        pCurrentHttpsResponse->pCallbacks->responseCompleteCallback( pCurrentHttpsResponse->pUserPrivData, pCurrentHttpsResponse, status, pCurrentHttpsResponse->status );
+    }
+
+    /* For a synchronous request release the semaphore. */
+    if( pCurrentHttpsResponse->isAsync == false )
+    {
+        IotSemaphore_Post( &( pCurrentHttpsResponse->respFinishedSem ) );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+static IotHttpsReturnCode_t _createHttpsConnection( IotHttpsConnectionHandle_t * pConnHandle,
+                                                    IotHttpsConnectionInfo_t * pConnInfo )
+{
+    HTTPS_FUNCTION_ENTRY( IOT_HTTPS_OK );
+
+    IotNetworkError_t networkStatus = IOT_NETWORK_SUCCESS;
+
+    /* The maximum string length of the ALPN protocols is configured in IOT_HTTPS_MAX_ALPN_PROTOCOLS_LENGTH.
+     * The +1 is for the NULL terminator needed by IotNetworkCredentials_t.pAlpnProtos. */
+    char pAlpnProtos[ IOT_HTTPS_MAX_ALPN_PROTOCOLS_LENGTH + 1 ] = { 0 };
+
+    /* The maximum string length of the Server host name is configured in IOT_HTTPS_MAX_HOST_NAME_LENGTH.
```

Comment:
There is a max value defined by the DNS standard. Can we just use that?

Reply:
It is defaulted to that in iot_https_internal.h:

```
#ifndef IOT_HTTPS_MAX_HOST_NAME_LENGTH
    #define IOT_HTTPS_MAX_HOST_NAME_LENGTH         ( 255 ) /* Per FQDN, the maximum host name length is 255 bytes. */
#endif
```